client-data/tools/grid/grid.js

```diff
@@ -0,0 +1,86 @@
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.  The code is distributed WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
+ *
+ * As additional permission under GNU GPL version 3 section 7, you
+ * may distribute non-source (e.g., minimized or compacted) forms of
+ * that code without the copy of the GNU GPL normally required by
+ * section 4, provided you include this license notice and a URL
+ * through which recipients can access the Corresponding Source.
+ *
+ * @licend
+ */
+
+(function grid() { //Code isolation
+
+    var index = 0; //grid off by default
+    var states = ["none", "url(#grid)", "url(#dots)"];
+
+    function toggleGrid(evt) {
+        index = (index+1) % states.length;
+        gridContainer.setAttributeNS(null, "fill", states[index]);
+    }
+
+    function createSVGElement(name, attrs) {
+        var elem = document.createElementNS("http://www.w3.org/2000/svg", name);
+        Object.keys(attrs).forEach(function(key, i) {
+            elem.setAttributeNS(null, key, attrs[key]);
+        });
+        return elem;
+    }
+
+    function init() {
+        // create patterns
+        // small (inner) grid
+        var smallGrid = createSVGElement("pattern", {id: "smallGrid", width: "30", height: "30", patternUnits: "userSpaceOnUse"});
+        smallGrid.appendChild(
+            createSVGElement("path", {d: "M 30 0 L 0 0 0 30", fill: "none", stroke: "gray", 'stroke-width': "0.5"})
+        );
+        // (outer) grid
+        var grid = createSVGElement("pattern", {id: "grid", width: "300", height: "300", patternUnits: "userSpaceOnUse"});
+        grid.appendChild(createSVGElement("rect", {width: "300", height: "300", fill: "url(#smallGrid)"}));
+        grid.appendChild(
+            createSVGElement("path", {d: "M 300 0 L 0 0 0 300", fill: "none", stroke: "gray", 'stroke-width': "1"})
+        );
+        // dots
+        var dots = createSVGElement("pattern", {id: "dots", width: "30", height: "30", x: "-10", y: "-10", patternUnits: "userSpaceOnUse"});
+        dots.appendChild(createSVGElement("circle", {fill: "gray", cx: "10", cy: "10", r: "2"}));
+
+        var defs = Tools.svg.getElementById("defs");
+        defs.appendChild(smallGrid);
+        defs.appendChild(grid);
+        defs.appendChild(dots);
+
+        // create grid container
+        var gridContainer = createSVGElement("rect", {id: "gridContainer", width: "100%", height: "100%", fill: states[index]});
+        Tools.svg.insertBefore(gridContainer, Tools.drawingArea);
+
+        return gridContainer;
+    }
+
+    var gridContainer = init();
+
+    Tools.add({ //The new tool
+        "name": "Grid",
+        "shortcut": "g",
+        "listeners": {},
+        "icon": "tools/grid/icon.svg",
+        "oneTouch": true,
+        "onstart": toggleGrid,
```

Comment:
What would you think of just letting the onstart function return whether the tool wants to be activated or not ? This would allow us not to have special code to handle `oneTouch` tools, and also let the tools themselves be more flexible about when the want/can be activated.

Reply:
I have encountered several issues trying to use the onstart function for this but it mainly boils down to this:
If we use onstart without changing anything in `board.js` the UI flickers since the tool will get changed anyway and we can only set a timeout to change it back.
If we instead modify call of the onstart function in the change handler to only change the tool if the onstart function returns a trueish value we will still have called the onquit function of the previous tool.
I don't see an elegant solution to this or anything that is less obstrusive than adding the oneTouch property.